Adding spectrogram, RMS, and SPL notebook

aditya-uw opened the pull request:
Hi @valentina-s, as Wu-Jung mentioned over in Zulip, I made a widget for playing around with spectrogram settings using an audio file uploaded into `BOAT_datasets`. I've added start time, duration, and NFFT sliders. You can find the notebook containing this widget under `boat_tutorials/exercises/BOAT_datasets_exercise.ipynb`

Will update when I add the calculation for RMS and SPL as well.

aditya-uw commented:
@leewujung I added the spectrogram, RMS voltage, SPL calculation, and even included power spectral density just in case. Let me know how I can clean this up if it needs to be more polished.

aditya-uw commented:
@leewujung check out `sound_reception_exercise.ipynb` as well when you have the time, I was able to make the kind of visual for 2 receivers that we were talking about. The beam pattern needs to be made still but wanted to get your quick thoughts on what I have.

leewujung commented:
This is very cool! Some suggestions:
- could you plot the sound source as a hollow circle or star or a different symbol? 
- could you separate the receiving blue and green in 1 subplot and the summed one in another subplot? maybe do upper 2/3 for subplot 1 and bottom 1/2 for subplot 2-- matplotlib gridspec may come in handy here! 
- offset blue and green vertically so it is easier to compare
- overall title for waveform plot: "Received voltage signal"
   - legend in subplot 1: rcvr #1/#2 according to the left
   - legend in subplot 2: summed voltage signal
- add frequency slider :D

aditya-uw commented:
@leewujung I addressed all suggestions about `sound_reception_exercise.ipynb` but I just wanted to check if you had any thoughts about `BOAT_datasets_exercise.ipynb`?

leewujung commented:
@aditya-uw I haven't looked at that! Will need a few hours before I get there. Feel free to go on to others widgets!

aditya-uw commented:
@leewujung sounds good, indeed that is what I'm doing!

aditya-uw commented:
@leewujung I have N-receivers simulation ready in `N_rcvr_sound_reception_exercise.ipynb`. Check it out when you can, I think I was confused on how the panels look less crowded when I increased the number of receivers (which increases number of received signals to plot). I just decided to go with whatever made sense so let me know your thoughts when you're able.

aditya-uw commented:
@leewujung This PR is ready to be merged. The additions are:
1) `exercises/BOAT_datasets_exercise.ipynb`:
   - Added cell to read in `.wav` file downloaded from Google Drive folder onto local machine
   - Added cell to play the `.wav` file contents in the form of a widget.
   - Added widget to plot a section of the `.wav` file as a time-series and as a spectrogram with variable parameters of duration and start time of section. Also includes variable NFFT parameter.
   - Added code to calculate the Vrms and SPL of a specific section.
   - Added method that converts Vrms to SPL using sensitivity, microphone gain, and preamp gain. Added inverse method that goes from SPL to Vrms.
   -Added widget at the end that plots time-series, spectrogram, FFT, and Welch's PSD for any section of the `.wav` file where start time, duration, and NFFT are variable.
2) `exercises/N_rcvr_sound_reception.ipynb`:
   - Has the widget where we compute the simulation of having an N-element vertical line array and a sound source that revolves around the origin. 
   - Each receiver is Ri and the distance between Ri to Source is computed and used to calculate the phase shift for the received signal at Ri. As source gets closer to any Ri, the received signal will move towards the left because it will be "received sooner". As source gets farther from any Ri, the received signal will move towards the right because it will be "received later". The plots are zoomed in on a larger signal duration so these signals will not appear truncated.
   - Received signals are generated for all angles as the first step and all plots only show a specific angle index of the larger matrix. Same applies for the summed signal which is shown below the subplot of received signals.
   - The summed signals for all angles is used to calculate the beam pattern. The Vrms is calculated for each summed signal and plotted along angles to get the beam pattern.
   
I added `N_rcvr_sound_reception.ipynb`'s widget into `acoustics/RL_03_directionality.ipynb` in the form of a fixed widget that only allows 2 receivers and then a widget that takes a variable 2 to 10 receivers.

I added the SPL to Voltage conversion widget from `BOAT_datasets_exercise.ipynb` into `acoustics/RL_02_sensitivity.ipynb` so people can sweep SPL, sensitivity, microphone gain, and preamp gain to get observed Vrms.

I know I wrote a lot but just wanted to get myself up-to-date on the additions so far.

leewujung commented:
Thanks @aditya-uw ! I'll merge this now so that we will be in sync.
